applications/MeshingApplication/custom_utilities/uniform_refine_utility.cpp

```diff
@@ -0,0 +1,680 @@
+            NodeType::Pointer middle_node = GetNodeInEdge(geom);
+
+            // Create the sub conditions
+            std::vector<NodeType::Pointer> sub_condition_nodes(2);
+            for (int position = 0; position < 2; position++)
+            {
+                sub_condition_nodes = GetSubLineNodes(position, geom, middle_node);
+                CreateCondition(p_condition, sub_condition_nodes, step_refine_level);
+            }
+        }
+        else
+        {
+            KRATOS_ERROR << "Your geometry contains " << geom.GetGeometryType() << " which cannot be refined" << std::endl;
+        }
+
+        // Once we have created all the sub conditions, the origin conditions must be deleted
+        p_condition->Set(TO_ERASE, true);
+    }
+
+    mrModelPart.RemoveConditionsFromAllLevels(TO_ERASE);
+
+}
+
+
+/// Create a middle node on an edge. If the node does not exist, it creates one
+template <unsigned int TDim>
+void UniformRefineUtility<TDim>::CreateNodeInEdge(
+    const EdgeType& rEdge,
+    const int& rRefinementLevel
+    )
+{
+    // Get the middle node key
+    std::pair<int, int> node_key;
+    node_key = std::minmax(rEdge(0)->Id(), rEdge(1)->Id());
+
+    // Check if the node is not yet created
+    auto search = mNodesMap.find(node_key);
+    if (search == mNodesMap.end() )
+    {
+        // Create the new node
+        double new_x = 0.5*rEdge(0)->X() + 0.5*rEdge(1)->X();
+        double new_y = 0.5*rEdge(0)->Y() + 0.5*rEdge(1)->Y();
+        double new_z = 0.5*rEdge(0)->Z() + 0.5*rEdge(1)->Z();
+        NodeType::Pointer middle_node = mrModelPart.CreateNewNode(++mLastNodeId, new_x, new_y, new_z);
+
+        // Store the node key in the map
+        mNodesMap[node_key] = middle_node->Id();
+
+        // interpolate the variables
+        CalculateNodalStepData(middle_node, rEdge(0), rEdge(1));
+
+        // Set the refinement level
+        int& this_node_level = middle_node->GetValue(REFINEMENT_LEVEL);
+        this_node_level = rRefinementLevel;
+
+        // Set the DoF's
+        for (typename NodeType::DofsContainerType::const_iterator it_dof = mDofs.begin(); it_dof != mDofs.end(); ++it_dof)
+            middle_node->pAddDof(*it_dof);
+
+        // Add the node to the sub model parts
+        int key0 = mNodesColorMap[rEdge(0)->Id()];
+        int key1 = mNodesColorMap[rEdge(1)->Id()];
+        int key = mSubModelPartsColors.IntersectKeys(key0, key1, mColors);
+        if (key != 0)  // NOTE: key==0 is the main model part
+        {
+            for (std::string sub_name : mColors[key])
+            {
+                ModelPart& sub_model_part = SubModelPartsListUtility::GetRecursiveSubModelPart(mrModelPart, sub_name);
+                sub_model_part.AddNode(middle_node);
+            }
+        }
+        mNodesColorMap[middle_node->Id()] = key;
+    }
+}
+
+
+/// Get the middle node on an edge
+template <unsigned int TDim>
+Node<3>::Pointer UniformRefineUtility<TDim>::GetNodeInEdge(const EdgeType& rEdge)
```

Comment:
Try to define a NodeType instead of use Node<3>, we will remove the template of the Node some day...